### PR TITLE
Benchcmp support in Makefile

### PR DESCRIPTION
**Pull request by glibsm**
Adds a few more targets to the makefile to easier generate the benchcmp after perf changes.

Workflow:

1. On master: `make benchbase`, that will create `.bench/old.txt`
2. Type type type perf changes
3. `make bench && make benchcmp` will show the diff before and after.

There are some guardrails about running `make benchbase` while not on master (should it be `origin/master?`), but can be overriden via `IGNORE_BASELINE_CHECK` (in case all you made so far are doc changes or something)

### CODE REVIEW

**Comment by glibsm** [comment]
@uber-go/service-framework

File: Makefile [hunk 1]
```diff
@@ -66,8 +66,37 @@ coveralls: $(COV_REPORT)
 
 .PHONY: bench
 BENCH ?= .
+BENCH_FILE ?= .bench/new.txt
 bench:
-	$(ECHO_V)$(foreach pkg,$(BENCH_PKGS),go test -bench=$(BENCH) -run="^$$" $(BENCH_FLAGS) $(pkg);)
+	$(ECHO_V)$(foreach pkg,$(BENCH_PKGS),go test -bench=. -run="^$$" $(BENCH_FLAGS) $(pkg) |\
```
Comment (sectioneight) [inline Makefile]: nit: space betweeen `|` and `\` to make it clear they're two separate operators

File: Makefile [hunk 2]
```diff
@@ -66,8 +66,37 @@ coveralls: $(COV_REPORT)
...
+		tee $(BENCH_FILE);)
+
+.PHONY: benchbase
+BASELINE_BENCH_FILE=.bench/old.txt
```
Comment (sectioneight) [inline Makefile]: you usually define the PHONY right above the target. so swp this line with line 74

File: Makefile [hunk 3]
```diff
@@ -66,8 +66,37 @@ coveralls: $(COV_REPORT)
...
+		exit 1; \
+	fi
+
+	@echo "$(LABEL_STYLE)Running baseline benchmark$(COLOR_RESET)"
+	@echo
+	$(ECHO_V)$(MAKE) bench BENCH_FILE=$(BASELINE_BENCH_FILE)
+
+.PHONY: benchcmp
+benchcmp:
+	$(ECHO_V)which benchcmp >/dev/null || go get -u golang.org/x/tools/cmd/benchcmp
```
Comment (sectioneight) [inline Makefile]: Do we want to consider vendoring this via glide to have consistent versions? Or is it stable

File: Makefile [hunk 4]
```diff
@@ -66,8 +66,37 @@ coveralls: $(COV_REPORT)
...
+	$(ECHO_V)test -s $(BASELINE_BENCH_FILE) ||\
```
Comment (sectioneight) [inline Makefile]: nit: space between `||` and `\` to make it obvious they're two separate operators

File: Makefile [hunk 5]
```diff
@@ -66,8 +66,37 @@ coveralls: $(COV_REPORT)
...
+		$(error "Must have a baseline bench file. *Check out master* and run `make benchbase`")
+	$(ECHO_V)test -s $(BENCH_FILE) || $(error "Must have a new benchmark file. Run `make bench`")
+	$(ECHO_V)benchcmp $(BASELINE_BENCH_FILE) $(BENCH_FILE)
+
+.PHONY: benchclean
+benchclean:
+	$(ECHO_V)if [ -a $(BASELINE_BENCH_FILE) ]; then rm $(BASELINE_BENCH_FILE); fi
```
Comment (sectioneight) [inline Makefile]: -a
file exists

This is identical in effect to -e. It has been "deprecated," [1] and its use is discouraged.

http://www.tldp.org/LDP/abs/html/fto.html

Also just do rm -f $(BASELINE_BENCH_FILE) and don't bother checking it first.

File: .build/colors.mk [hunk 6]
```diff
@@ -13,3 +13,6 @@ endif
 
 LABEL_STYLE = "$(COLOR_BOLD)$(COLOR_COMMAND)"
 ERROR_STYLE = "$(COLOR_BOLD)$(COLOR_ERROR)"
+
+label = echo $(COLOR_BOLD)$(COLOR_COMMAND)$(1)$(COLOR_NORMAL)
```
Comment (sectioneight) [inline .build/colors.mk]: WOAH. TILabout `$(1)`. I've been wanting this forever!

File: Makefile (hunk 3 above)
Comment (glibsm) [inline Makefile]: I think it's pretty stable, we can vendor it in later if we want

Comment (sectioneight) [inline Makefile]: Agreed. I'm full-on YAGNI and did this with thriftrw, which bit us eventually but would rather start everything simple and occasionally have to revisit than overdo it from the beginning.

**Comment by sectioneight** [comment]
This is really, really cool Glib. Nice work.

**Pull request merged**